File: python-backend/app/services/model_service.py
from __future__ import annotations

from app.config import ModelConfig, Settings
from app.errors import NotFoundError
from app.models import ModelResponse


class ModelService:
    def __init__(self, settings: Settings) -> None:
        self._settings = settings

    def list_models(self) -> list[ModelResponse]:
        enabled_models = [item for item in self._settings.models if item.enabled]
        ordered_models = sorted(
            enabled_models,
            key=lambda item: (item.id != self._settings.default_model_id, item.provider == "local", item.name),
        )
        return [
            ModelResponse(
                id=item.id,
                name=item.name,
                provider=item.provider,
                modelName=item.model_name,
                enabled=item.enabled,
            )
            for item in ordered_models
        ]

    def get_model(self, model_id: str) -> ModelConfig:
        resolved_id = model_id or self._settings.default_model_id
        for item in self._settings.models:
            if item.id == resolved_id and item.enabled:
                return item
        raise NotFoundError("模型不存在或未启用")

File: backend/src/main/java/com/example/assistant/application/model/ModelService.java
package com.example.assistant.application.model;

import com.example.assistant.application.chat.BusinessException;
import com.example.assistant.application.chat.ErrorCode;
import com.example.assistant.domain.model.AiModel;
import com.example.assistant.infrastructure.config.AssistantProperties;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
public class ModelService {

    private final String defaultModelId;
    private final List<AiModel> enabledModels;
    private final Map<String, AiModel> enabledModelMap;

    public ModelService(AssistantProperties properties) {
        this.defaultModelId = properties.defaultModelId().strip();
        this.enabledModels = properties.models().stream()
                .filter(AssistantProperties.ModelConfig::enabled)
                .map(this::toDomain)
                .sorted((left, right) -> Boolean.compare(!left.id().equals(defaultModelId), !right.id().equals(defaultModelId)))
                .toList();
        if (enabledModels.isEmpty()) {
            throw new IllegalStateException("assistant.models must contain at least one enabled model");
        }
        this.enabledModelMap = enabledModels.stream()
                .collect(Collectors.toUnmodifiableMap(AiModel::id, Function.identity()));
        if (!enabledModelMap.containsKey(defaultModelId)) {
            throw new IllegalStateException("assistant.default-model-id must reference an enabled model");
        }
    }

    public Flux<AiModel> listEnabledModels() {
        return Flux.fromIterable(enabledModels);
    }

    public Mono<AiModel> getEnabledModel(String modelId) {
        String selectedModelId = modelId == null || modelId.isBlank() ? defaultModelId : modelId.strip();
        return Mono.justOrEmpty(enabledModelMap.get(selectedModelId))
                .switchIfEmpty(Mono.error(new BusinessException(ErrorCode.MODEL_NOT_FOUND)));
    }

    private AiModel toDomain(AssistantProperties.ModelConfig config) {
        return new AiModel(
                config.id().strip(),
                config.name().strip(),
                config.provider().strip(),
                config.modelName().strip(),
                StringUtils.hasText(config.baseUrl()) ? config.baseUrl().strip() : "",
                StringUtils.hasText(config.apiKey()) ? config.apiKey().strip() : "",
                config.enabled()
        );
    }
}
